src/libstd/thread/mod.rs

```diff
@@ -1036,15 +1037,15 @@ pub fn park_timeout(dur: Duration) {
 /// [`Thread`]: ../../std/thread/struct.Thread.html
 #[stable(feature = "thread_id", since = "1.19.0")]
 #[derive(Eq, PartialEq, Clone, Copy, Hash, Debug)]
-pub struct ThreadId(u64);
+pub struct ThreadId(NonZeroU64);
```

Comment:
If you want to allow more aggressive niche optimizations you can also reserve a few (thousand?) values at the end of the value range via `#[rustc_layout_scalar_valid_range_end(0xFFFF_FFFF_FFFF_0000)]` (that will require unsafe code though)

Reply:
I’m a bit wary of adding more use of `rustc_layout_scalar_valid_range_end` when there seems to be no plan to ever stabilize it. And I don’t know how common are cases when this would make a difference over non-zero.